src/librustc/hir/lowering/expr.rs

```diff
@@ -514,7 +514,7 @@ impl LoweringContext<'_> {
 
     /// Desugar `<expr>.await` into:
     /// ```rust
-    /// match <expr> {
+    /// match ::std::future::IntoFuture::into_future(<expr>) {
```

Comment:
Oh, I think I know why that test changed, but also, that it's entirely avoidable.
What seems to be happening is something very silly which is that if `<expr>` has any `.await`s itself, the `IntoFuture:: into_future` value is considered to be "live at a suspend point".

I don't see why MIR building would generate that instead of keeping a constant operand around, but if nobody has the time to fix this for all `f(g().await)` (i.e. make it equivalent to `{let x = g().await; f(x)}`) you can do a manual transformation here and leave a FIXME comment pointing to an issue to clean it up.
(cc @matthewjasper @Zoxc for the MIR side of this)

The manually transformed form would be something like `match { let x = <expr>; IntoFuture:: into_future(x) } {...}`.

Reply:
Just remembered that typeck collects a list of types without being aware of whether they might show up in the MIR.

I think we can whitelist call expressions where the callee is a path to a function or method (not based on type, since you can put a function in a local variable).

Reply:
> The manually transformed form would be something like `match { let x = <expr>; IntoFuture:: into_future(x) } {...}`.

I don't grasp the *why*, but should I try this? Local testing is hard, but I can make this change and remove the `type_level_limit` attribute from that test, push and see if CI passes.

Reply:
The why is easy: `f(g().await)` right now keeps a temporary `f` around across the suspend point for the inner await - or at least `rustc_typeck` assumes that happens.

It might help if you think of this as `call(f, call(g).await)`, where `f` is necessarily evaluated *before* the suspend, but the outer `call` is necessarily evaluated *after* the suspend, requiring the result of evaluating `f` to stick around (which also means its type ends up in the generator type).

Reply:
The MIR should be irrelevant since the type we're instantiating is decided by typeck. That information is in the `GeneratorInterior`, built by the code in this file: https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/generator_interior.rs#L169

I'm not sure exactly what special case @eddyb had in mind, but I think it would go in one of the methods of the `InteriorVisitor`.

Reply:
Hopefully we can rely on https://github.com/rust-lang/rust/pull/65743 for the fix.